### PR TITLE
test(filestorage): add community ObjectStore extension-path template

### LINKED ISSUE #4912: [IMPROVEMENT] C-SR-17: Tests: `register_object_store` characterization
Supports Python API / repeat usage — repeat-usage.
Audience: community
Kind (maintainer): n/a
Task id: C-SR-17
Code PRs: TDD; AAA (Arrange→Act→Assert); clear names; in-tree patterns; SOLID.
Not: `core/` product edits.

---

| Field | Value |
|-------|-------|
| **Audience** | Community |
| **Level** | Junior→Middle |
| **Complexity** | S |
| **Good first issue** | Yes |
| **North star** | Daily use (extensibility) |

**Summary:** Add characterization / example tests that a tiny fake `ObjectStore`
can `register_object_store("fake", …)`, build via registry, and run a pull/push
against the engine with temp dirs — so community providers have a template.
**TDD:** write the test that fails if registration is broken, then keep fakes only.

**Do not:** Add a real cloud SDK. Do not edit `core/`. Do not change production
providers beyond what registration requires (prefer test-only register/unregister).

**Done when:** [ ] Test file documents the extension path
[ ] Tests use AAA (Arrange → Act → Assert)
[ ] Register / unregister cleaned up in fixture teardown
[ ] `uv run pytest` on that file green

**Files:** `tests/` under filestorage (≤2 files)

---

### PR DESCRIPTION
**Pull request by arnavpopley**
Fixes #4912

#### Describe the changes you have made in this PR -

Added `tests/filestorage/test_community_provider_template.py`: a worked example for community contributors adding a new remote-sync backend. Documents the whole contract - implement the four `ObjectStore` methods, scope keys through `config.key_for()`/prefix-stripping like every built-in provider, record a real per-object write time, protect shared state since `put_object` is called concurrently, then `register_object_store`.

- `_FakeObjectStore`: takes `RemoteSyncConfig` and a shared `bucket` dict (one bucket, many clients), scopes keys by prefix, records write time once at `put_object`, guards every read/write with a lock.
- `fake_provider` fixture: registers/unregisters "fake" via `yield` teardown.
- `test_a_registered_fake_provider_is_built_via_the_registry`: registry resolves the factory.
- `test_push_then_pull_round_trips_a_file_through_the_fake_store`: real `push()`/`pull()` against the engine with `tmp_path`, no mocks.
- `test_objects_are_scoped_under_the_configured_prefix`: two stores, same bucket, different prefixes, cannot see each other's objects.
- `test_put_object_records_a_stable_write_time_not_a_fresh_one_per_listing`: two listings return an identical `last_modified`.
- `test_put_object_is_safe_under_the_engines_real_concurrent_push`: a real concurrent push (barrier-forced overlap, same technique as `test_push_concurrency.py`) proves no write is lost.

One file, no production code touched.

**Revision history:** two earlier versions were closed after Greptile review (#5064: missing prefix scoping and stale timestamps; #5065: undocumented, unlocked concurrency contract on `put_object`). All three are fixed here, each with a test that fails against the unfixed implementation.

### Demo/Screenshot for feature changes and bug fixes -

```
pytest tests/filestorage/test_community_provider_template.py -v -> 5 passed in 1.44s
pytest tests/filestorage/ -q -> 275 passed, 1 warning in 3.93s
ruff check / ruff format --check / mypy -> all clean
```

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

Read every built-in provider and `ports.py`'s `ObjectStore.put_object` docstring to confirm the full contract: prefix scoping, real write timestamps, and the "called concurrently, do not mutate shared state unprotected" note. Two review rounds caught two real gaps in what the template taught. Each is fixed with the minimal correct pattern (a lock around the shared dict, matching real providers' thread-safe SDK clients) and proven with a targeted test, not just described in prose - the concurrency test in particular uses a barrier rather than a plain file count, since a race in an unlocked plain dict is not reliably reproducible under CPython's GIL without forcing genuine overlap.

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

### CODE REVIEW

**Comment by arnavpopley** [comment]
@greptile review

File: tests/filestorage/test_community_provider_template.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+    for i in range(cap):
+        (root / f"file{i}.jsonl").write_text(f'{{"n": {i}}}\n', encoding="utf-8")
+    roots = (SyncRoot(name=SyncRootName.SESSIONS, path=root),)
+
+    # Act
+    report = push(store, roots=roots, max_parallel_uploads=cap)
+
+    # Assert
+    assert len(report.uploaded) == cap
+    assert len(store.list_objects("")) == cap
```
[the remark this answers is not in the document]
Comment (arnavpopley) [reply tests/filestorage/test_community_provider_template.py]: Resolved

**Comment by muddlebee** [comment]
@arnavpopley thank you for the PRs, but as those are labelled "good first issues" for first time contributors only

**Comment by arnavpopley** [comment]
@muddlebee I'd already spoken to Vaibhav before; he told me to finish up the unassigned good first issues
